### PR TITLE
chore: implement experimental ESM stub/spy for Vite

### PR DESCRIPTION
**Pull request by lmiller1990**
- Closes https://github.com/cypress-io/cypress-internal/issues/564

### Additional details

First release of our Vite plugin to facilitate spy/mock on ES modules.

I made this `chore` since

1) It's still a POC and has known issues
2) I want it to say below v1 until we've tested it more with some of the pilot users

Not sure the best way to handle this, but `chore` should only increment the patch number.

### Steps to test

This is based on the PROPOSAL section from the tech brief: https://github.com/cypress-io/engineering-documentation/tree/main/tech-briefs/ct-esm-mocks#proposal. That explains what's going on here in detail.

It's complicated, don't feel you like need 100% understand it all when you review it. 

Testing: Run the test suite! Or make a new Vite project that uses ES Modules and write some tests using `cy.spy` and `cy.stub`. The things in our documentation for [`cy.stub()`](https://docs.cypress.io/api/commands/stub) and [`cy.spy()`](https://docs.cypress.io/api/commands/spy) should work correctly with ES modules. There's a bunch of tests in this PR demonstrating it works (mostly, sans the caveats in the README).

There are some caveats, I wrote those in the README.

### How has the user experience changed?

You can spy/mock ES modules (both static and dynamic imports).

### PR Tasks

- [x] Have tests been added/updated?
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

File: npm/vite-plugin-cypress-esm/client/moduleCache.js [hunk 1]
```diff
@@ -0,0 +1,145 @@
+const __cypressModuleCache = new Map()
```
Comment (lmiller1990) [inline npm/vite-plugin-cypress-esm/client/moduleCache.js:1]: This map is going to become really huge over time, I think we probably want to clear it - maybe on `Cypress:on('test:suite:start')` or something.

There's a bunch of improvements that can likely be made, these will come over time - the goal is to get something that works as expected in the hands of users asap, and we can benchmark/optimize etc as we go.

File: npm/vite-plugin-cypress-esm/README.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
+# @cypress/vite-plugin-cypress-esm
+
+A Vite plugin that intercepts and rewrites ES module imports. It wraps them in a `Proxy`, allowing for instrumentation by libraries such as Sinon. Useful for test runners using Vite as a dev server.
+
+> **Note:** This package is a pre-release and is not yet stable. There are likely to be bugs and edge cases. Please report and bugs [here](https://github.com/cypress-io/cypress/issues/new?labels=npm:%20@cypress/vite-plugin-cypress-esm)
```
Comment (mike-plummer) [inline npm/vite-plugin-cypress-esm/README.md]:
```suggestion
> **Note:** This package is a pre-release and is not yet stable. There are likely to be bugs and edge cases. Please report any bugs [here](https://github.com/cypress-io/cypress/issues/new?labels=npm:%20@cypress/vite-plugin-cypress-esm)
```

File: npm/vite-plugin-cypress-esm/README.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+Run Cypress with `DEBUG=cypress:vite-plugin-cypress-esm`. You will get logs in the terminal, for the code transformation, and in the browser console, for intercepting and wrapping the modules in a Proxy. 
+## Compatibility
+
+| @cypress/vite-plugin-mock-esm | cypress |
+| ------------------------ | ------- |
+| >= v1                    | >= v12  |
+
+## Usage
+
+This is tampers with your ES modules to make them mutable (thus compatible with methods like `cy.spy()` and `cy.stub()` that require modifying otherwise sealed objects) you probably only want to apply this during your Cypress tests. One way to do so would be in `cypress.config`:
```
Comment (mike-plummer) [inline npm/vite-plugin-cypress-esm/README.md]:
```suggestion
This plugin rewrites the ES modules served by Vite to make them mutable and therefore compatible with methods like [`cy.spy()`](https://docs.cypress.io/api/commands/spy) and [`cy.stub()`](https://docs.cypress.io/api/commands/stub) that require modifying otherwise-sealed objects. Since this is a testing-specific plugin it is recommended to apply it your Vite config only when running your Cypress tests. One way to do so would be in `cypress.config`:
```

File: npm/vite-plugin-cypress-esm/README.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
+# @cypress/vite-plugin-cypress-esm
+
+A Vite plugin that intercepts and rewrites ES module imports. It wraps them in a `Proxy`, allowing for instrumentation by libraries such as Sinon. Useful for test runners using Vite as a dev server.
```
Comment (mike-plummer) [inline npm/vite-plugin-cypress-esm/README.md]:
```suggestion
A Vite plugin that intercepts and rewrites ES module imports within [Cypress component tests](https://docs.cypress.io/guides/component-testing/overview). The [ESM specification](https://tc39.es/ecma262/#sec-modules) generates modules that are "sealed", requiring the runtime (the browser) to prevent any alteration to the module namespace. While this has security and performance benefits, it prevents use of mocking libraries which would need to replace namespace members. This plugin wraps modules in a special [`Proxy`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) implementation, allowing for instrumentation by libraries such as Sinon.
```

File: npm/vite-plugin-cypress-esm/README.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+import defaultExport3, * as name2 from "./module";
+```
+
+An alterntive would simply be to split the import syntax over two lines:
+
+```js
+import defaultExport3 from "./module";
+import * as name2 from "./module";
+```
+
```
Comment (mike-plummer) [inline npm/vite-plugin-cypress-esm/README.md:99]: Worth adding a mention of potential perf/memory impacts for *huge* test suites?

File: npm/vite-plugin-cypress-esm/client/moduleCache.js [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+      if (typeof value === 'function') {
+        const stack = new Error().stack
+
+        if (stack.includes('Sandbox.spy')) {
+          return value
+        }
+
+        return proxies[prop]
+      }
```
Comment (mike-plummer) [inline npm/vite-plugin-cypress-esm/client/moduleCache.js]: Some comments here for posterity would probably be good

```suggestion
      if (typeof value === 'function') {
        // Check to see if this retrieval is coming from a sinon `spy` creation
        // If so, we want to supply the 'true' function rather than our proxied version
        // so the spy can call through to the real implementation
        const stack = new Error().stack

        if (stack.includes('Sandbox.spy')) {
          return value
        }

        // Otherwise, return our proxied function implementation
        return proxies[prop]
      }
```

File: npm/vite-plugin-cypress-esm/client/moduleCache.js [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+  console.log(`[cypress:vite-plugin-proxify-esm]: ${msg}`)
+}
+
+function cacheAndProxifyModule (id, module, debug) {
+  if (__cypressModuleCache.has(module)) {
+    return __cypressModuleCache.get(module)
+  }
+
+  if (debug) {
+    log(`🔨 creating proxy module for ${id}`)
```
Comment (mike-plummer) [inline npm/vite-plugin-cypress-esm/client/moduleCache.js]: Ha, love the emojis

File: npm/vite-plugin-cypress-esm/client/moduleCache.js [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+      return target[prop]
+    },
+    set (obj, prop, value) {
+      target[prop] = value
+
+      if (typeof value === 'function' && !(prop in proxies)) {
+        proxies[prop] = function (...params) {
+          return target[prop].apply(this, params)
+        }
```
Comment (mike-plummer) [inline npm/vite-plugin-cypress-esm/client/moduleCache.js:126]: nitpick: This duplicates the function wrapping we do on line 45, could pull out a `wrapFunction` helper just to make sure they stay aligned

Comment (mike-plummer) [reply npm/vite-plugin-cypress-esm/client/moduleCache.js:126]: Now that I think about it I can't remember why I found this block necessary. Theoretically all functions *should* have been proxied up front 🤔

File: npm/vite-plugin-cypress-esm/package.json [hunk 9]
```diff
@@ -0,0 +1,41 @@
+{
+  "name": "@cypress/vite-plugin-cypress-esm",
```
Comment (mike-plummer) [inline npm/vite-plugin-cypress-esm/package.json:2]: Thoughts on the double "cypress" in the name here?

File: npm/vite-plugin-cypress-esm/package.json [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    "lint": "eslint --ext .js,.ts,.json, ."
+  },
+  "dependencies": {
+    "debug": "^4.3.4",
+    "semver": "^7.3.2"
+  },
+  "devDependencies": {
+    "@vitejs/plugin-react": "1.3.1",
+    "react": "16.8.6",
+    "react-dom": "16.8.6",
```
Comment (mike-plummer) [inline npm/vite-plugin-cypress-esm/package.json:23]: Pretty old versions here, any reason why or just copied from elsewhere?

File: npm/vite-plugin-cypress-esm/package.json [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    "build-prod": "tsc || echo 'built, with type errors'",
+    "check-ts": "tsc --noEmit",
+    "cypress:open": "node ../../scripts/cypress open --project . --component",
+    "test": "node ../../scripts/cypress run --project . --browser chrome --component",
+    "watch": "tsc -w",
+    "lint": "eslint --ext .js,.ts,.json, ."
+  },
+  "dependencies": {
+    "debug": "^4.3.4",
+    "semver": "^7.3.2"
```
Comment (mike-plummer) [inline npm/vite-plugin-cypress-esm/package.json]: Is `semver` actually used anywhere? I don't see it

File: npm/vite-plugin-cypress-esm/package.json [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    "dist",
+    "client",
+    "index.d.ts"
+  ],
+  "license": "MIT",
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/cypress-io/cypress.git"
+  },
+  "homepage": "https://github.com/cypress-io/cypress/tree/develop/npm/vite-plugin-proxify-esm#readme",
```
Comment (mike-plummer) [inline npm/vite-plugin-cypress-esm/package.json]: Looks like the plugin name changed, need to update link?

File: npm/vite-plugin-cypress-esm/package.json [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+  "bugs": "https://github.com/cypress-io/cypress/issues/new?template=1-bug-report.md",
```
Comment (mike-plummer) [inline npm/vite-plugin-cypress-esm/package.json]: nit: Can we add the `labels=npm:%20@cypress/vite-plugin-cypress-esm`label via this link?

File: npm/vite-plugin-cypress-esm/src/index.ts [hunk 14]
```diff
@@ -0,0 +1,123 @@
+import debugFn from 'debug'
+import type { Plugin } from 'vite'
+import fs from 'fs'
+import path from 'path'
+const debug = debugFn('cypress:vite-plugin-mock-esm')
```
Comment (mike-plummer) [inline npm/vite-plugin-cypress-esm/src/index.ts:6]: Align debug name with new plugin name?

File: npm/vite-plugin-cypress-esm/src/index.ts [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+   *
+   * Given:
+   *   const m = import("./mod_1")
+   *   const m = await import("lodash")
+   *   import("./mod_2").then(mod => mod)
+   *
+   * Returns:
+   *   const m = cypressModule(import("./mod_1"))
+   *   const m = await cypressModule(import("lodash"))
+   *   cypressModule(import("./mod_2")).then(mod => mod)
```
Comment (mike-plummer) [inline npm/vite-plugin-cypress-esm/src/index.ts]:
```suggestion
   * Transform dynamic imports to use a runtime that turns the ES modules into proxies by injecting
   * a `__cypressDynamicModule` runtime.
   *
   * `__cypressDynamicModule` returns a Promise that resolves to the original import which has been
   * "proxified" to side step the fact ES Modules are immutable and sealed.
   *
   * Examples - https://regex101.com/r/Ic1OHA/1
   *
   * Given:
   *   const m = import("./mod_1")
   *   const m = await import("lodash")
   *   import("./mod_2").then(mod => mod)
   *
   * Returns:
   *   const m = __cypressDynamicModule(import("./mod_1"))
   *   const m = await __cypressDynamicModule(import("lodash"))
   *   __cypressDynamicModule(import("./mod_2")).then(mod => mod)
```

File: npm/vite-plugin-cypress-esm/src/index.ts [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+    const moduleIdentifier = moduleId.replace(/[^a-zA-Z\d]/g, '_')
+
+    const importRegex = /import (.+?) from (.*)/g
+
+    return code.replace(
+      importRegex,
+      (match, importVars: string, importTarget: string) => {
+        debug(`Mapping import ${counter + 1} in module ${moduleId}`)
+
+        let replacement = `import * as cypress_${moduleIdentifier}_${++counter} from ${importTarget};`
```
Comment (mike-plummer) [inline npm/vite-plugin-cypress-esm/src/index.ts]: It hadn't really registered with me that doing an `import *` would destroy any tree-shaking, but I guess there isn't much of that done in dev mode so maybe not a huge impact?

File: npm/vite-plugin-cypress-esm/src/index.ts [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+
+    return c
+  }
+
+  return {
+    name: 'cypress:mocks',
+    enforce: 'post',
+    transform (code, id, options) {
+      // Process all files to remap imports
+      // TODO: Restrict to JS files only? Any potential for .mjs etc?
```
Comment (mike-plummer) [inline npm/vite-plugin-cypress-esm/src/index.ts:153]: Any reason not to lock this down to `.*js` files? Minor perf bump if we bypass processing json, images, etc

**Review by mike-plummer — CHANGES_REQUESTED** [review]
Looks amazing, going to do some testing today but leaving a few comments on the code now. Not sure what's going on with the V8 snapshots and the `yarn.lock` file, don't think any of those changes should be necessary. Maybe your workspace is in a weird state?

File: npm/vite-plugin-cypress-esm/README.md (hunk 5 above)
Comment (lmiller1990) [reply npm/vite-plugin-cypress-esm/README.md:99]: I realized since we've got a new tab per spec it should be cleaned up in between specs for free, I don't expect this to be an issue unless people use the Single Tab Run Mode (experimental).

File: npm/vite-plugin-cypress-esm/client/moduleCache.js (hunk 8 above)
Comment (lmiller1990) [reply npm/vite-plugin-cypress-esm/client/moduleCache.js:126]: I tried to make this refactor, it's quite awkward and I think it's a little less clear to read. Any recommendation for how you think it could be better written?

File: npm/vite-plugin-cypress-esm/package.json (hunk 9 above)
Comment (lmiller1990) [reply npm/vite-plugin-cypress-esm/package.json:2]: It's not ideal, but if you look at it from a "package owned by an org" without the second `cypress`, it's just `vite-plugin-esm` -  which it certainly *could* be, but right now it's Cypress specific.

Any ideas?

Or, we can pivot `__cypressModule` to `__proxyModule` etc. and make it agnostic, and drop the second `cypress`?

File: npm/vite-plugin-cypress-esm/package.json (hunk 11 above)
Comment (lmiller1990) [reply npm/vite-plugin-cypress-esm/package.json]: I don't think so, I think it can be removed. I'll  remove it.

File: npm/vite-plugin-cypress-esm/package.json (hunk 12 above)
Comment (lmiller1990) [reply npm/vite-plugin-cypress-esm/package.json]: done

File: npm/vite-plugin-cypress-esm/src/index.ts (hunk 14 above)
Comment (lmiller1990) [reply npm/vite-plugin-cypress-esm/src/index.ts:6]: done

File: npm/vite-plugin-cypress-esm/src/index.ts (hunk 16 above)
Comment (lmiller1990) [reply npm/vite-plugin-cypress-esm/src/index.ts]: I have no idea about perf yet, we should definitely do some benchmarks once we get all the bugs sorted out.

**Comment by lmiller1990** [comment]
I found a bug. Comment: https://github.com/muratkeremozcan/tour-of-heroes-react-vite-cypress-ts/pull/103#issuecomment-1515728750

Something in our logic is preventing prototype based inheritance working correctly, I think... I reproduced it and added a test using the `react-query` library. Details below, but before that, I tried side stepping the one module (`react-query`) that was giving me issues:

```js
  // Check if it is the offending module - don't proxify
  if ('useQueryErrorResetBoundary' in module) {
    return module
  }
```

This works - test runs fine. Instead of proxifying every module, we could do a pre-scan of the spec and only apply the proxy logic to things matching `cy.stub()` or `cy.spy()`, maybe using a regexp like `cy\.(spy|stub)\((.+?)`. This would not work if the user wants to stub/spy a library that doesn't work with our Proxy logic - we don't really know how common those libraries are yet.

Onto my debugging:

I tracked the code down. The error is on the line `this.options = this.client.defaultMutationOptions(options)
` , which is called [here](https://github.com/TanStack/query/blob/317fe0822606590c08a9e3ce0952ab785ae34912/packages/query-core/src/mutationObserver.ts#L69). It says `defaultMutationOptions` is not a function. This is supposed to be inherited from the `clientQuery` class, it is defined [here](https://github.com/TanStack/query/blob/317fe0822606590c08a9e3ce0952ab785ae34912/packages/query-core/src/queryClient.ts#L776).

I added a minimal test using `class` and basic inheritance - works great. Eg:

```js
class Animal {
  voice : string = 'Hello'
  greet() {
    return this.voice
  }
}

export class Dog extends Animal {
  constructor(voice: string) {
    super()
    this.voice = voice
  }
}

new Dog('bark').greet() // works great
```

The issue is, I think the type of code generated when `react-query` is compiled. It spits out:

```js
import _inheritsLoose from "@babel/runtime/helpers/esm/inheritsLoose";
import { Subscribable } from './subscribable';

export var MutationObserver = /*#__PURE__*/function (_Subscribable) {
  _inheritsLoose(MutationObserver, _Subscribable);

  function MutationObserver(client, options) {
    var _this;

    _this = _Subscribable.call(this) || this;
    _this.client = client;

    _this.setOptions(options);

    return _this;
  }

  var _proto = MutationObserver.prototype;

  _proto.setOptions = function setOptions(options) {
    this.options = this.client.defaultMutationOptions(options);
  };

  return MutationObserver;
}(Subscribable);
```

Which seems to not play nicely with our proxy code. After some debugging, the `MutationObserver` is supposed to have a `Subscribable` prototype, but after the proxy code runs, it does not have the correct prototype, it has `Object`. 

I need to debug more, but it's not clear if 1) we've hit an actual limitation of what proxy can do or 2) if we can fix this. I suspect we should not be using `Object` but `Reflect` in general, which has more powerful introspection APIs.

File: npm/vite-plugin-cypress-esm/package.json (hunk 9 above)
Comment (mike-plummer) [reply npm/vite-plugin-cypress-esm/package.json:2]: Leaving the second "cypress" works for me, better to eliminate any potential confusion - it's really not at all general-purpose at this point (has some unique behaviors to account for Sinon that probably wouldn't work with other libs)

File: npm/vite-plugin-cypress-esm/client/moduleCache.js (hunk 8 above)
Comment (mike-plummer) [reply npm/vite-plugin-cypress-esm/client/moduleCache.js:126]: Just wondering now that we have more custom logic in this area if we'll need to apply it (eg constructor handling) in other places we're building function wrappers - there's like 4 places we're generating wrappers, hard to keep track of which ones need to account for what.

File: npm/vite-plugin-cypress-esm/src/index.ts (hunk 17 above)
Comment (mike-plummer) [reply npm/vite-plugin-cypress-esm/src/index.ts:153]: Ugh, just realized we ship "*.vue", "*.svelte", etc for the iframe source and they need to be transformed as well since they can contain JS. Guess we'd either have to whitelist and risk missing some weird extensions or blacklist any that are obviously not JS that could be hefty and slow down the server with unnecessary regexing (PNG, JPG, JSON, etc). Not worth it for now

Comment (lmiller1990) [reply npm/vite-plugin-cypress-esm/src/index.ts:153]: Yeah, let's deal with this at some point in the future. For now I think as long as we work well with React that's probably fine!

File: npm/vite-plugin-cypress-esm/package.json (hunk 10 above)
Comment (lmiller1990) [reply npm/vite-plugin-cypress-esm/package.json:23]: I didn't want to mess with the yarn.lock or the v8 snapshots, so I just chose versions of libraries we've already got in the monorepo.

File: npm/vite-plugin-cypress-esm/client/moduleCache.js (hunk 8 above)
Comment (lmiller1990) [reply npm/vite-plugin-cypress-esm/client/moduleCache.js:126]: We can probably rejigger this a bit now - we've got decent enough test coverage.

Would also be good to get this merged prior to the sprint ending - I could go either way. Hard to say how much more we will work on this in the near future, given the caveats we've run into (like the reference vs proxy comparison that might be unsolvable).

File: npm/vite-plugin-cypress-esm/package.json (hunk 9 above)
Comment (lmiller1990) [reply npm/vite-plugin-cypress-esm/package.json:2]: Sounds good to me, can reconsider if/when we hit 1.0.0 for this.

**Review by mike-plummer — APPROVED** [review]

File: npm/vite-plugin-cypress-esm/cypress/component/dynamicImport.cy.ts [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+  it('uses real implementation', () => {
+    expect(_.add(1, 2)).to.eq(3)
+  })
+
+  it('mocks', () => {
+    cy.stub(_, 'add').callsFake(function multiply (a: number, b: number) {
+      return a * b
+    })
+
+    // Plot twist - add actually does multiplication!
```
Comment (marktnoonan) [inline npm/vite-plugin-cypress-esm/cypress/component/dynamicImport.cy.ts:15]: Plot twist 😆

File: npm/vite-plugin-cypress-esm/cypress/component/edgeCases.cy.tsx [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+import * as Mod from './fixtures/class'
+import * as Foo from './fixtures/Foo'
+import numbers from './fixtures/defaultExportArray'
+import { App } from './fixtures/reactQuery'
+import { letters } from './fixtures/namedExportArray'
+import { BrowserRouter, Routes, Route } from 'react-router-dom'
+const About = lazy(() => import('./fixtures/exportDefaultConst'))
+
+describe('edge cases', () => {
+  describe('class with constructor', () => {
```
Comment (marktnoonan) [inline npm/vite-plugin-cypress-esm/cypress/component/edgeCases.cy.tsx:14]: Nit: Always a bit odd to me when there's a nested describe at the top, since Cypress runs this at the bottom, after the `it` blocks below this. We could move this to the bottom, or wrap the tests below in their own describe block. Purely so it's easier to open the file and have the order in the command log be similar to the source order.

File: npm/vite-plugin-cypress-esm/cypress/component/fixtures/class.ts [hunk 20]
```diff
@@ -0,0 +1,21 @@
+export class Foo {
```
Comment (marktnoonan) [inline npm/vite-plugin-cypress-esm/cypress/component/fixtures/class.ts:1]: I find it hard to follow Foo/Bar/Bax examples, they don't give much context for what I should expect the classes to do and I lose track of them quite quickly. Not a blocker ofc.

File: npm/vite-plugin-cypress-esm/README.md [hunk 21]
```diff
@@ -0,0 +1,123 @@
+# @cypress/vite-plugin-cypress-esm
+
+A Vite plugin that intercepts and rewrites ES module imports within [Cypress component tests](https://docs.cypress.io/guides/component-testing/overview). The [ESM specification](https://tc39.es/ecma262/#sec-modules) generates modules that are "sealed", requiring the runtime (the browser) to prevent any alteration to the module namespace. While this has security and performance benefits, it prevents use of mocking libraries which would need to replace namespace members. This plugin wraps modules in a special [`Proxy`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) implementation, allowing for instrumentation by libraries such as Sinon.
+
+> **Note:** This package is a pre-release and is not yet stable. There are likely to be bugs and edge cases. Please report any bugs [here](https://github.com/cypress-io/cypress/issues/new?labels=npm:%20@cypress/vite-plugin-cypress-esm)
```
Comment (marktnoonan) [inline npm/vite-plugin-cypress-esm/README.md]: Might be a good place to link to the release stages docs.

```suggestion
> **Note:** This package is a pre-release alpha and is not yet stable. There are likely to be bugs and edge cases. Please report any bugs [here](https://github.com/cypress-io/cypress/issues/new?labels=npm:%20@cypress/vite-plugin-cypress-esm). [Learn more about Cypress release stages](https://docs.cypress.io/guides/references/release-stages#Alpha) and expectations around stability.
```

File: npm/vite-plugin-cypress-esm/README.md [hunk 22]
```diff
@@ -0,0 +1,123 @@
...
+
+You can also use a glob, which uses [`picomatch`](https://github.com/micromatch/picomatch) internally:
+
+```ts
+CypressEsm({
+  ignoreList: ['*react*']
+})
+```
+
+React is known to have some conflicts with the Proxy implementation. You probably don't want to stub your UI library anyway, so it's a good idea to add it to the `ignoreList`.
```
Comment (marktnoonan) [inline npm/vite-plugin-cypress-esm/README.md]: At a glance this looks like you might not be able to use this for stubbing within React apps. How about this alteration:

```suggestion
React is known to have some conflicts with the Proxy implementation that cause problems stubbing internal React functionality. Since it is unlikely you want to stub parts of React itself, it's a good idea to add it to the `ignoreList`.
```

File: npm/vite-plugin-cypress-esm/README.md [hunk 23]
```diff
@@ -0,0 +1,123 @@
...
+
+## Known Issues
+
+### Import Syntax
+
+All known [import syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import) is supported, however there may edge cases that have not been identified
```
Comment (marktnoonan) [inline npm/vite-plugin-cypress-esm/README.md]:
```suggestion
All known [import syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import) is supported, however there may edge cases that have not been identified.
```

File: npm/vite-plugin-cypress-esm/README.md [hunk 24]
```diff
@@ -0,0 +1,123 @@
...
+
+In this example, `bar(foo)` is passing a reference to `mod_1.foo`, where `mod_1` is a module wrapped in a `Proxy`. In the original `mod_1.js`, the reference to `foo` is the original, unwrapped `foo`, so the comparison return `false`. This may cause issues in some libraries, such as React Router when lazy loading routes. You can add modules to `ignoreList` to work around this issue.
+
+### Sinon compatibility
+
+This plugin is designed to work with [Sinon](https://sinonjs.org/) since that is what Cypress uses internally for `cy.stub` and `cy.spy` - attempting to utilize other stubbing/mocking libraries or directly mutating modules is not a supported use case and will likely not work as expected.
+
+## Troubleshooting
+
+This is an **_Alpha_** release, meaning there a very likely bugs in the implementation and it is expected that you will encounter issues. We appreciate any bug reports once you have performed the troubleshooting process below
```
Comment (marktnoonan) [inline npm/vite-plugin-cypress-esm/README.md]:
```suggestion
This is an **_Alpha_** release, meaning there a very likely bugs in the implementation and it is expected that you will encounter issues. We appreciate any bug reports once you have performed the troubleshooting process below.
```

File: npm/vite-plugin-cypress-esm/src/index.ts [hunk 25]
```diff
@@ -0,0 +1,184 @@
...
+        html: transformedHtml,
+        tags: [
+          // Pull in Cypress module cache script
+          {
+            tag: 'script',
+            attrs: {
+              type: 'module',
+            },
+            // eslint-disable-next-line no-restricted-syntax
+            children: fs.readFileSync(MODULE_CACHE_FILEPATH, 'utf-8'),
```
Comment (marktnoonan) [inline npm/vite-plugin-cypress-esm/src/index.ts:178]: Is there a reason we ignore this rule here? We say `Synchronous fs calls should not be used in Cypress. Use an async API instead.` so might worth commenting `using readFileSync because x`.

**Review by marktnoonan — APPROVED** [review]
This worked well for me in testing, I've left a few minor, non blocking comments in the review and some suggestions for the readme.

**Review by astone123 — APPROVED** [review]
Nice, hopefully we can move this into `vite-dev-server` once we get more feedback from users 👍🏻
